File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        nums.sort()

        maxDiff = float('inf')
        ans = nums[0] + nums[1] + nums[2]

        for i in range(n - 2):
            left = i + 1
            right = n - 1

            while left < right:

                total = nums[i] + nums[left] + nums[right]
                difference = abs(target - total)

                if difference < maxDiff:
                    maxDiff = difference
                    ans = total

                if total == target:
                    return total

                elif total < target:
                    left += 1

                else:
                    right -= 1

        return ans

File: 0016-3sum-closest/0016-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int n = nums.size();
        int ans;
        int maxDiff = INT_MAX;
        sort(nums.begin(),nums.end());


        for (int i =0;i<n-2;i++){
            int left =i+1;
            int right = n-1;

            while(left<right){
                int sum = nums[i]+nums[left]+nums[right];
                int difference = abs(target-sum);
                if(difference<maxDiff){
                    maxDiff = difference;
                    ans = sum;
                }
                if(sum==target){
                    return sum;
                }
                if(sum<target){
                    left++;
                }
                if(sum>target){
                    right--;
                }
            }

            
        }
        return ans;

    }
};